File: services/mirrors.py
import json
import socket
import requests
from threading import Lock
from utils.logger import logger
from services.database import db_lock, get_db_connection
from utils.config import get_config

# In-memory cache of DB mirrors
# Structure: {'name': {'urls': [...], 'status': 'approved'}}
MIRRORS_CACHE = {} 
mirrors_lock = Lock()

def load_mirrors_from_db():
    """Load mirrors from database"""
    global MIRRORS_CACHE
    try:
        with db_lock:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute('SELECT name, urls, status FROM mirrors')
            rows = cursor.fetchall()
            
            with mirrors_lock:
                MIRRORS_CACHE.clear()
                for row in rows:
                    try:
                        MIRRORS_CACHE[row['name']] = {
                            'urls': json.loads(row['urls']),
                            'status': row['status']
                        }
                    except:
                        pass
            
            conn.close()
            logger.info(f"Loaded {len(MIRRORS_CACHE)} mirrors from database")
    except Exception as e:
        logger.error(f"Error loading mirrors from DB: {e}")

def is_self(host):
    """Check if the host refers to this server"""
    # Strip port
    hostname = host.split(':')[0]
    
    if hostname in ['localhost', '127.0.0.1', '::1', '0.0.0.0']:
        return True
        
    try:
        host_ip = socket.gethostbyname(hostname)
        
        # Get local IPs
        local_ips = set()
        local_hostname = socket.gethostname()
        try:
            local_ips.add(socket.gethostbyname(local_hostname))
        except:
            pass
            
        try:
            for info in socket.getaddrinfo(local_hostname, None):
                if info[4][0]:
                    local_ips.add(info[4][0])
        except:
            pass
            
        if host_ip in local_ips:
            return True
    except:
        pass
        
    return False

def validate_mirror(url):
    """Check if the mirror URL is reachable"""
    try:
        # Use HEAD request to check if it exists
        resp = requests.head(url, timeout=5, allow_redirects=True)
        return resp.status_code < 400
    except:
        return False

def save_mirror_to_db(name, urls, status='pending'):
    """Save a new dynamic mirror to the database"""
    
    # Validation: Check if self
    if is_self(name):
        logger.warning(f"Skipping self-referencing mirror: {name}")
        return False
        
    # Validation: Check reachability
    valid_urls = []
    for url in urls:
        if validate_mirror(url):
            valid_urls.append(url)
            
    if not valid_urls:
        logger.warning(f"No valid/reachable URLs for mirror: {name}")
        return False
        
    try:
        with db_lock:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute('INSERT OR REPLACE INTO mirrors (name, urls, status) VALUES (?, ?, ?)', 
                          (name, json.dumps(valid_urls), status))
            conn.commit()
            conn.close()
            
        with mirrors_lock:
            MIRRORS_CACHE[name] = {
                'urls': valid_urls,
                'status': status
            }
            
        logger.info(f"Saved mirror: {name} -> {valid_urls} ({status})")
        return True
    except Exception as e:
        logger.error(f"Error saving mirror to DB: {e}")
        return False

def update_mirror(name, urls=None, status=None):
    """Update an existing mirror's URLs and/or status"""
    try:
        with mirrors_lock:
            if name not in MIRRORS_CACHE:
                return False
            current_data = MIRRORS_CACHE[name]
            
        new_urls = urls if urls is not None else current_data['urls']
        new_status = status if status is not None else current_data['status']
        
        if status is not None and status not in ['approved', 'pending', 'blacklisted']:
            return False

        # If URLs are being updated, validate them
        if urls is not None:
            valid_urls = []
            for url in urls:
                if validate_mirror(url):
                    valid_urls.append(url)
            if not valid_urls:
                logger.warning(f"No valid URLs provided for update: {name}")
                return False
            new_urls = valid_urls

        with db_lock:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute('UPDATE mirrors SET urls = ?, status = ? WHERE name = ?', 
                          (json.dumps(new_urls), new_status, name))
            if cursor.rowcount == 0:
                return False
            conn.commit()
            conn.close()
            
        with mirrors_lock:
            MIRRORS_CACHE[name] = {
                'urls': new_urls,
                'status': new_status
            }
                
        logger.info(f"Updated mirror: {name} -> {new_urls} ({new_status})")
        return True
    except Exception as e:
        logger.error(f"Error updating mirror: {e}")
        return False

def delete_mirror(name):
    """Delete a mirror from the database"""
    try:
        with db_lock:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute('DELETE FROM mirrors WHERE name = ?', (name,))
            conn.commit()
            conn.close()
            
        with mirrors_lock:
            if name in MIRRORS_CACHE:
                del MIRRORS_CACHE[name]
                
        logger.info(f"Deleted mirror: {name}")
        return True
    except Exception as e:
        logger.error(f"Error deleting mirror: {e}")
        return False

def get_all_mirrors():
    """Get only APPROVED mirrors for proxy usage"""
    # We no longer use config.json for mirrors, only DB
    approved_mirrors = {}
    with mirrors_lock:
        for name, data in MIRRORS_CACHE.items():
            if data['status'] == 'approved':
                approved_mirrors[name] = data['urls']
    return approved_mirrors

def get_mirrors_management():
    """Get ALL mirrors for management UI"""
    with mirrors_lock:
        return MIRRORS_CACHE.copy()

def get_upstream_key(distro, package_path):
    """Determine which upstream mirror to use based on path patterns"""
    path_lower = package_path.lower()
    mirrors = get_all_mirrors()
    
    # Check for security
    if 'security' in path_lower:
        sec_key = f"{distro}-security"
        if sec_key in mirrors:
            return sec_key
            
    return distro

File: internal/mirrors/mirrors.go
package mirrors

import (
	"encoding/json"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"

	"apt-cache-proxy/internal/database"
	"apt-cache-proxy/internal/logger"
)

type Mirror struct {
	URLs   []string `json:"urls"`
	Status string   `json:"status"`
}

var (
	mirrorsCache map[string]Mirror
	mu           sync.RWMutex
)

func init() {
	mirrorsCache = make(map[string]Mirror)
}

// LoadFromDB loads mirrors from database
func LoadFromDB() error {
	db := database.Get()
	log := logger.Get()
	
	rows, err := db.Query("SELECT name, urls, status FROM mirrors")
	if err != nil {
		return err
	}
	defer rows.Close()
	
	mu.Lock()
	defer mu.Unlock()
	
	mirrorsCache = make(map[string]Mirror)
	
	for rows.Next() {
		var name, urlsJSON, status string
		if err := rows.Scan(&name, &urlsJSON, &status); err != nil {
			continue
		}
		
		var urls []string
		if err := json.Unmarshal([]byte(urlsJSON), &urls); err != nil {
			continue
		}
		
		mirrorsCache[name] = Mirror{
			URLs:   urls,
			Status: status,
		}
	}
	
	log.Infof("Loaded %d mirrors from database", len(mirrorsCache))
	return nil
}

// GetAll returns all approved mirrors
func GetAll() map[string][]string {
	mu.RLock()
	defer mu.RUnlock()
	
	result := make(map[string][]string)
	for name, mirror := range mirrorsCache {
		if mirror.Status == "approved" {
			result[name] = mirror.URLs
		}
	}
	return result
}

// GetAllWithStatus returns all mirrors with their status (for admin)
func GetAllWithStatus() map[string]Mirror {
	mu.RLock()
	defer mu.RUnlock()
	
	result := make(map[string]Mirror)
	for name, mirror := range mirrorsCache {
		result[name] = mirror
	}
	return result
}

// GetUpstreamKey determines the upstream key from distro and path
func GetUpstreamKey(distro, pkgPath string) string {
	// Handle Ubuntu releases/pockets
	if strings.HasPrefix(distro, "ubuntu") || strings.Contains(distro, "noble") ||
		strings.Contains(distro, "jammy") || strings.Contains(distro, "focal") {
		return "ubuntu"
	}
	
	// Handle Debian releases
	if strings.HasPrefix(distro, "debian") || strings.Contains(distro, "bookworm") ||
		strings.Contains(distro, "bullseye") || strings.Contains(distro, "buster") {
		return "debian"
	}
	
	return distro
}

// Save saves a mirror to database
func Save(name string, urls []string, status string) error {
	// Validate: check for self-reference
	if isSelf(name) {
		log := logger.Get()
		log.Warnf("Skipping self-referencing mirror: %s", name)
		return nil
	}
	
	// Validate URLs
	validURLs := []string{}
	for _, url := range urls {
		if validateMirror(url) {
			validURLs = append(validURLs, url)
		}
	}
	
	if len(validURLs) == 0 {
		return nil
	}
	
	urlsJSON, err := json.Marshal(validURLs)
	if err != nil {
		return err
	}
	
	db := database.Get()
	_, err = db.Exec("INSERT OR REPLACE INTO mirrors (name, urls, status) VALUES (?, ?, ?)",
		name, string(urlsJSON), status)
	if err != nil {
		return err
	}
	
	// Update cache
	mu.Lock()
	mirrorsCache[name] = Mirror{
		URLs:   validURLs,
		Status: status,
	}
	mu.Unlock()
	
	log := logger.Get()
	log.Infof("Saved mirror: %s (%d URLs, status: %s)", name, len(validURLs), status)
	return nil
}

// Update updates a mirror's URLs or status
func Update(name string, urls []string, status string) error {
	db := database.Get()
	
	if urls != nil {
		urlsJSON, err := json.Marshal(urls)
		if err != nil {
			return err
		}
		_, err = db.Exec("UPDATE mirrors SET urls = ? WHERE name = ?", string(urlsJSON), name)
		if err != nil {
			return err
		}
	}
	
	if status != "" {
		_, err := db.Exec("UPDATE mirrors SET status = ? WHERE name = ?", status, name)
		if err != nil {
			return err
		}
	}
	
	// Reload from DB
	return LoadFromDB()
}

// Delete deletes a mirror
func Delete(name string) error {
	db := database.Get()
	_, err := db.Exec("DELETE FROM mirrors WHERE name = ?", name)
	if err != nil {
		return err
	}
	
	mu.Lock()
	delete(mirrorsCache, name)
	mu.Unlock()
	
	log := logger.Get()
	log.Infof("Deleted mirror: %s", name)
	return nil
}

func isSelf(host string) bool {
	hostname := strings.Split(host, ":")[0]
	
	if hostname == "localhost" || hostname == "127.0.0.1" || hostname == "::1" || hostname == "0.0.0.0" {
		return true
	}
	
	// Check if it resolves to a local IP
	ips, err := net.LookupIP(hostname)
	if err != nil {
		return false
	}
	
	// Get local IPs
	localIPs := getLocalIPs()
	
	for _, ip := range ips {
		for _, localIP := range localIPs {
			if ip.Equal(localIP) {
				return true
			}
		}
	}
	
	return false
}

func getLocalIPs() []net.IP {
	var ips []net.IP
	
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return ips
	}
	
	for _, addr := range addrs {
		if ipnet, ok := addr.(*net.IPNet); ok {
			ips = append(ips, ipnet.IP)
		}
	}
	
	return ips
}

func validateMirror(url string) bool {
	client := &http.Client{
		Timeout: 5 * time.Second,
	}
	
	resp, err := client.Head(url)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	
	return resp.StatusCode < 400
}
